Modules/Core/Common/include/itkImageSink.h

```diff
@@ -0,0 +1,189 @@
+/*=========================================================================
+ *
+ *  Copyright Insight Software Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkImageSink_h
+#define itkImageSink_h
+
+#include "itkStreamingProcessObject.h"
+#include "itkImage.h"
+#include "itkImageRegionSplitterBase.h"
+#include "itkImageRegionSplitterSlowDimension.h"
+#include "itkImageToImageFilterCommon.h"
+
+namespace itk
+{
+
+/** \class ImageSink
+ *
+ * ImageSink is the base class for process objects which consume image
+ * data. This class defaults to having at least one input of the
+ * templated image type. The framework enables derived algorithms to
+ * stream the input image as it's being consumed by the algorithm.
+ *
+ * The framework provides multi-threading of the streamed image regions.
+ * The input image's pipeline is updated multiple time with the
+ * streaming requested regions, then the  fulfilled requested region
+ * are split again for multi-threading.
+ *
```

Comment:
* By default, the NumberOfStreamDivisions is just 1 (no streaming),
 * derived implementations must change the access specification for
 * this method to be public to expose the streaming feature.